### PR TITLE
 Backport CFG simplification pass from XLA backend

### PR DESCRIPTION
**Pull request by Keno**
Right now we don't do too much CFG simplification (esp since we turned off doing so for constant folded branches). This isn't too much of a problem, because LLVM is very good at cutting down any excess basic block we happen to emit.

However, for non-SSA backends, excess CFG can be a significant problem that the backend may not be able to optimize away (even if it's trivial at the Julia IR level). Plus it's annoying for humans to read. The XLA backend had a simple CFG simplification pass. Backport this pass to Base, so it can live alongside the code it depends on (it has a fairly close dependency on the details of the CFG and IncrementalCompact). As it stands, I don't think it's useful to have this pass in the default compiler pipeline (both because LLVM can handle it easily and because our round-trip to statement based representations cleans up some of this), but I do think it's useful interactively and for non-standard compiler backends. We should re-evaluate whether to put this in the standard compiler pipeline once we re-enable the CFG transformations of constant folded conditions. If that ends up leaving a lot of basic block chains around, it might yet be worth putting this in.

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
What's the cost of turning it on? If we're going to have it, it would be better to have it turned on so that it actually gets used and tested in the normal course of things. Does it slow things down?

**Comment by Keno** [comment]
> What's the cost of turning it on? If we're going to have it, it would be better to have it turned on so that it actually gets used and tested in the normal course of things. Does it slow things down?

It's probably not super expensive, but the real problem is that it doesn't buy us anything. It's exercised in the test suite, which is the primary reason I want it here so changes to the underlying code notice if they break this pass. I think it's totally fine to have passes that are provided for external users only (after all most of the high level functionality in Base is there to be used by external users). If you'd prefer, we can also put this in a new CompilerTools stdlib or similar.

**Comment by StefanKarpinski** [comment]
Have you tried enabling it and seeing the effect? I have to wonder if doing a little cheap early simplification doesn't sometimes save time in the later passes. I'm fine with leaving it here just as long as it's tested so it doesn't bitrot.

File: base/compiler/ssair/ir.jl [hunk 1]
```diff
@@ -514,7 +515,7 @@ mutable struct IncrementalCompact
         new_new_nodes = NewNode[]
         pending_nodes = NewNode[]
         pending_perm = Int[]
-        return new(code, result, result_types, result_lines, result_flags, result_bbs, ssa_rename, bb_rename, used_ssas, late_fixup, perm, 1,
+        return new(code, result, result_types, result_lines, result_flags, result_bbs, ssa_rename, bb_rename, bb_rename, used_ssas, late_fixup, perm, 1,
             new_new_nodes, pending_nodes, pending_perm,
```
Comment (jrevels) [inline base/compiler/ssair/ir.jl]: I guess we set these to different arrays anyway in `cfg_simplify`, but maybe we should just initialize them as two separate arrays here? just so downstream code doesn't get tricked by this aliasing

Comment (Keno) [reply base/compiler/ssair/ir.jl]: It intentionally uses the same array. The array is read only unless cfg transforms are enabled.

**Comment by Keno** [comment]
There isn't really a good place to put it in our pipeline. In the XLA pipeline it's after an additional domsort pass at the end of our regular pass list.

**Comment by Keno** [comment]
Updated with the latest version from the XLA backend.
Changes:
- Fix a bug where a chain ending in a conditional would end up with the wrong fall through block (added test for this case also).
- Drop unreachable blocks as we go along (the next compact would have done so anyway, but it's nice not to need it).

**Comment by Keno** [comment]
Rebased because I needed it on another project. @vchuravy do you care about using this in Cthulhu?

**Comment by vchuravy** [comment]
> do you care about using this in Cthulhu?

Yes. Always.

**Comment by Keno** [comment]
Alright, in that case it probably makes sense to try to merge this (I need to keep separate branches for everything else anyway, so I don't care as much).

**Pull request closed without merging**